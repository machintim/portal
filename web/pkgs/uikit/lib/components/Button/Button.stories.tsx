import type { Meta, StoryObj } from '@storybook/react';

import Button from './Button';

const meta = {
  title: 'Components/Work in progress/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Default button',
  },
};

export const Primary: Story = {
  args: {
    variant: 'primary',
    children: 'Primary button',
  },
};

export const Secondary: Story = {
  args: {
    variant: 'secondary',
    children: 'Secondary button',
  },
};

export const Tertiary: Story = {
  args: {
    variant: 'tertiary',
    children: 'Tertiary button',
  },
};

export const Destructive: Story = {
  name: 'Destructive primary',
  args: {
    destructive: true,
    children: 'Destructive button',
  },
};

export const DestructiveSecondary: Story = {
  name: 'Destructive secondary',
  args: {
    variant: 'secondary',
    destructive: true,
    children: 'Destructive secondary button',
  },
};

export const Small: Story = {
  name: 'Small primary',
  args: {
    size: 'small',
    children: 'Small button',
  },
};

export const Large: Story = {
  name: 'Large primary',
  args: {
    size: 'large',
    children: 'Large button',
  },
};

export const SmallSecondary: Story = {
  name: 'Small secondary',
  args: {
    size: 'small',
    variant: 'secondary',
    children: 'Small secondary button',
  },
};

export const LargeSecondary: Story = {
  name: 'Large secondary',
  args: {
    size: 'large',
    variant: 'secondary',
    children: 'Large secondary button',
  },
};

export const Disabled: Story = {
  name: 'Primary disabled',
  args: {
    disabled: true,
    children: 'Disabled button',
  },
};

export const DisabledSecondary: Story = {
  name: 'Secondary disabled',
  args: {
    variant: 'secondary',
    disabled: true,
    children: 'Disabled secondary button',
  },
};

export const DisabledTertiary: Story = {
  name: 'Tertiary disabled',
  args: {
    variant: 'tertiary',
    disabled: true,
    children: 'Disabled tertiary button',
  },
};
