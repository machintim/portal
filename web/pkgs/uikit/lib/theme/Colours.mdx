import { Meta, Title, Subtitle, ColorPalette, ColorItem } from '@storybook/addon-docs/blocks';
import theme from './defaultTheme';

<Meta title="theme/Colours" />
<Title>Colours explorer</Title>
<Subtitle>Colours approved for use in the UCL Design System</Subtitle> 

More information on the use of colour in the UCL Design System <a href="https://www.figma.com/design/8Sm5PxWOWJYpYXAzhRzUzt/UCL-Design-System-UI-Kit?node-id=1-287" target="_blank">can be found on Figma</a>. 

The colours populating this page are imported from the <a href="https://github.com/ucl-isd/uikit-react/blob/main/lib/theme/defaultTheme.ts#L5" target="_blank">theme file</a>. 

Below are all colours per category, one category per line: 

<ColorPalette>
{
  Object.entries(theme.color).map(([key, value]) => {
    return (
      <ColorItem key={key} title={key.charAt(0).toUpperCase() + key.slice(1)} colors={value} />
    );
  })
}
</ColorPalette>

Below is a breakdown of each colour category. These are the same colours as above, but grouped by category, one colour per line: 

{Object.entries(theme.color).map(([key, value]) => {
  return (
    <div key={key}>
      <h2>{key.charAt(0).toUpperCase() + key.slice(1)}</h2>

      <ColorPalette>

      {Object.entries(value).map(([colorKey, colorValue]) => {
        return (
          <ColorItem key={colorKey} title={colorKey} colors={{ [colorKey]: colorValue }} />
        );
      })}

      {console.log(key, value)}

      </ColorPalette>
    </div>
  );
})}
