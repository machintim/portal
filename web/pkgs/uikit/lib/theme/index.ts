export { default as theme } from './defaultTheme';
export type {
  ThemeType,
  DisplayColoursType,
  BlackAndWhiteColoursType,
} from './defaultTheme';
export {
  default as useTheme,
  ThemeContextProvider,
} from './useTheme';
