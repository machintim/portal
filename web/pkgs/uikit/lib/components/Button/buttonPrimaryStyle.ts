import { css, cx } from '@emotion/css';
import { ThemeType } from '../../theme';
import { ButtonProps } from './Button';

const buttonPrimaryStyle = (
  theme: ThemeType,
  destructive: ButtonProps['destructive'],
  disabled: ButtonProps['disabled']
) => {
  const {
    color: { system, interaction },
    boxShadow,
  } = theme;

  const bgColour = destructive ? system.red70 : interaction.blue70;
  const bgHoverColour = destructive ? system.red100 : interaction.blue80;
  const bgActiveColour = destructive ? system.red70 : interaction.blue70;
  const boxShadowActive = destructive
    ? boxShadow.destructiveActive
    : boxShadow.active;

  const primaryBaseStyle = css`
    border: none;
    background-color: ${bgColour};
    color: ${theme.color.neutral.white};
    font-weight: ${theme.font.weight.bold};
  `;

  const interactiveStyle = css`
    cursor: pointer;

    &:hover {
      background-color: ${bgHoverColour};
    }

    &:active {
      background-color: ${bgActiveColour};
      box-shadow: ${boxShadowActive};
    }
  `;

  const disabledStyle = css`
    border-color: ${theme.color.text.disabled};
    background-color: ${theme.color.link.disabled};
    color: ${theme.color.text.disabledOnBg};
    cursor: not-allowed;
  `;

  return cx(
    primaryBaseStyle,
    !disabled && interactiveStyle,
    disabled && disabledStyle
  );
};

export default buttonPrimaryStyle;
