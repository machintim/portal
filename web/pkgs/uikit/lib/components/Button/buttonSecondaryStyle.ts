import { css, cx } from '@emotion/css';
import { ThemeType } from '../../theme';
import { ButtonProps } from './Button';

const buttonSecondaryStyle = (
  theme: ThemeType,
  destructive: ButtonProps['destructive'],
  disabled: ButtonProps['disabled']
) => {
  const {
    color: { system, interaction },
    boxShadow,
  } = theme;

  const colour = destructive ? system.red70 : interaction.blue70;
  const hoverColour = destructive ? system.red100 : interaction.blue80;
  const activeColour = destructive ? system.red70 : interaction.blue70;
  const boxShadowActive = destructive
    ? boxShadow.destructiveActive
    : boxShadow.active;

  const secondaryBaseStyle = css`
    background-color: ${theme.color.neutral.white};
    color: ${colour};
    border: ${theme.border.b1} solid ${colour};
    font-weight: ${theme.font.weight.regular};
  `;

  const interactiveStyle = css`
    cursor: pointer;

    &:hover {
      background-color: ${theme.color.neutral.white};
      border-color: ${hoverColour};
      color: ${hoverColour};
    }

    &:active {
      background-color: ${theme.color.neutral.white};
      color: ${activeColour};
      border-color: ${activeColour};
      box-shadow: ${boxShadowActive};
    }
  `;

  const disabledStyle = css`
    border-color: ${theme.color.text.disabled};
    color: ${theme.color.text.disabled};
    cursor: not-allowed;
  `;

  return cx(
    secondaryBaseStyle,
    !disabled && interactiveStyle,
    disabled && disabledStyle
  );
};

export default buttonSecondaryStyle;
