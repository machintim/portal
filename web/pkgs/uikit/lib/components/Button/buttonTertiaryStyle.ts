import { css, cx } from '@emotion/css';
import { ThemeType } from '../../theme';
import { ButtonProps } from './Button';

const buttonTertiaryStyle = (
  theme: ThemeType,
  disabled: ButtonProps['disabled']
) => {
  const {
    color: { interaction },
  } = theme;

  const colour = interaction.blue70;
  const hoverColour = interaction.blue80;
  const activeColour = interaction.blue70;

  const tertiaryBaseStyle = css`
    border: none;
    background-color: transparent;
    color: ${colour};
    font-weight: ${theme.font.weight.bold};
  `;

  const interactiveStyle = css`
    cursor: pointer;

    &:hover {
      color: ${hoverColour};
    }

    &:active {
      color: ${activeColour};
    }
  `;

  const disabledStyle = css`
    color: ${theme.color.text.disabled};
    cursor: not-allowed;
  `;

  return cx(
    tertiaryBaseStyle,
    !disabled && interactiveStyle,
    disabled && disabledStyle
  );
};

export default buttonTertiaryStyle;
