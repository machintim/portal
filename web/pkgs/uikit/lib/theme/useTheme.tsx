// We will probably move away from context/provider to store/hook for accessing theme.

import {
  createContext,
  useState,
  useContext,
  ReactNode,
} from 'react';

import theme, { ThemeType } from './defaultTheme';

const useThemeState = () => useState(theme as ThemeType);

const ThemeContext = createContext<ReturnType<
  typeof useThemeState
> | null>(null);

export const ThemeContextProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => (
  <ThemeContext.Provider value={useThemeState()}>
    {children}
  </ThemeContext.Provider>
);

const useTheme = () => {
  const value = useContext(ThemeContext);
  if (value === null) {
    throw new Error('Please add ThemeContextProvider');
  }
  return value;
};

export default useTheme;
