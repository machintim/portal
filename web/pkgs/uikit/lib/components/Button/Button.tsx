import { memo, ButtonHTMLAttributes, forwardRef } from 'react';
import { css, cx } from '@emotion/css';
import useTheme from '../../theme/useTheme';
import buttonPrimaryStyle from './buttonPrimaryStyle';
import buttonSecondaryStyle from './buttonSecondaryStyle';
import buttonTertiaryStyle from './buttonTertiaryStyle';

export const NAME = 'ucl-uikit-button';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'primary' | 'secondary' | 'tertiary';
  destructive?: boolean;
  size?: 'large' | 'default' | 'small';
  disabled?: boolean;
  testId?: string;
}

export type Ref = HTMLButtonElement;

const Button = forwardRef<Ref, ButtonProps>(
  (
    {
      variant = 'primary',
      destructive = false,
      size = 'default',
      disabled = false,
      testId = NAME,
      children,
      className,
      ...props
    },
    ref
  ) => {
    if (variant === 'tertiary' && destructive) {
      console.warn(
        'Button: tertiary and destructive props cannot be used together'
      );
    }

    const [theme] = useTheme();

    const baseStyle = css`
      font-family: ${theme.font.family.primary};
      font-size: ${theme.font.size.f16};
      outline: none;

      &:focus-visible {
        box-shadow: ${theme.boxShadow.focus};
      }
    `;

    const sizeSmallStyle = css`
      height: 40px;
      padding: 0 ${theme.padding.p16};
    `;

    const sizeDefaultStyle = css`
      height: 48px;
      padding: 0 ${theme.padding.p24};
    `;

    const sizeLargeStyle = css`
      height: 56px;
      padding: 0 ${theme.padding.p24};
      font-size: ${theme.font.size.f18};
    `;

    const variantStyle = {
      primary: buttonPrimaryStyle(theme, destructive, disabled),
      secondary: buttonSecondaryStyle(theme, destructive, disabled),
      tertiary: buttonTertiaryStyle(theme, disabled),
    }[variant];

    const style = cx(
      NAME,
      baseStyle,
      size === 'small' && sizeSmallStyle,
      size === 'default' && sizeDefaultStyle,
      size === 'large' && sizeLargeStyle,
      variantStyle,
      className
    );

    return (
      <button
        ref={ref}
        disabled={disabled}
        className={style}
        data-testid={testId}
        {...props}
      >
        {children}
      </button>
    );
  }
);

export default memo(Button);
