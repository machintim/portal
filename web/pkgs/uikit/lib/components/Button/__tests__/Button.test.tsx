import { describe, expect, test, vitest } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from '../Button';
import { ThemeContextProvider } from '../../../theme/useTheme';

describe('Button', () => {
  // Snapshot tests

  test('snapshot: no props', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Button />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  test('snapshot: variant prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Button variant='secondary' />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  test('snapshot: testId prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Button testId='test123' />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  // Interaction tests

  test('test ID: default', () => {
    render(
      <ThemeContextProvider>
        <Button />
      </ThemeContextProvider>
    );
    const button = screen.getByTestId('ucl-uikit-button');
    expect(button).toBeDefined();
  });

  test('test ID: custom', () => {
    render(
      <ThemeContextProvider>
        <Button testId='custom-test-id' />
      </ThemeContextProvider>
    );
    const button = screen.getByTestId('custom-test-id');
    expect(button).toBeDefined();
  });

  test('button click', async () => {
    const user = userEvent.setup();
    const testClick = vitest.fn();
    render(
      <ThemeContextProvider>
        <Button onClick={testClick} />
      </ThemeContextProvider>
    );

    const button = screen.getByRole('button') as HTMLButtonElement;
    expect(testClick).not.toBeCalled();
    await user.click(button);
    expect(testClick).toBeCalled();
  });

  test('button click disabled', async () => {
    const user = userEvent.setup();
    const testClick = vitest.fn();
    render(
      <ThemeContextProvider>
        <Button
          onClick={testClick}
          disabled
        />
      </ThemeContextProvider>
    );

    const button = screen.getByRole('button') as HTMLButtonElement;
    expect(testClick).not.toBeCalled();
    await user.click(button);
    expect(testClick).not.toBeCalled();
  });
});
