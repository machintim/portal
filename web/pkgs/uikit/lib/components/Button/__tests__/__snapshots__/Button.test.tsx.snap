// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > snapshot: no props 1`] = `
<button
  class="ucl-uikit-button css-1q5jb4"
  data-testid="ucl-uikit-button"
/>
`;

exports[`Button > snapshot: testId prop 1`] = `
<button
  class="ucl-uikit-button css-1q5jb4"
  data-testid="test123"
/>
`;

exports[`Button > snapshot: variant prop 1`] = `
<button
  class="ucl-uikit-button css-an7wpu"
  data-testid="ucl-uikit-button"
/>
`;
